STM32L4: CAN support

<!--
====================================
IF YOUR ISSUE IS RELATED TO SECURITY
====================================
please submit it to the security mailing-list security@riot-os.org.

If your issue is a question related to the usage of RIOT, please submit it to
the user mailing-list users@riot-os.org or to the developer mailing-list
devel@riot-os.org.
-->

#### Description
I'm trying to use a Nucleo-l476rg with CAN transceiver [MCP2551](https://www.microchip.com/wwwproducts/en/en010405):

- As reported by nucleo's MCU [datasheet]( https://www.st.com/resource/en/datasheet/stm32l476je.pdf), it has one CAN.
- STM32 can driver should be added since 74ddf56171d28858
- I take a look to [test/conn_can](https://github.com/RIOT-OS/RIOT/tree/master/tests/conn_can)

TODO:

1. Add the missing feature to the board profile:
```
diff --git a/boards/nucleo-l476rg/Makefile.features b/boards/nucleo-l476rg/Makefile.features
index b61dd0e29b48..eb91dc27bc15 100644
--- a/boards/nucleo-l476rg/Makefile.features
+++ b/boards/nucleo-l476rg/Makefile.features
@@ -3,6 +3,7 @@ CPU_MODEL = stm32l476rg
 
 # Put defined MCU peripherals here (in alphabetical order)
 FEATURES_PROVIDED += periph_adc
+FEATURES_PROVIDED += periph_can
 FEATURES_PROVIDED += periph_dma
 FEATURES_PROVIDED += periph_i2c
 FEATURES_PROVIDED += periph_pwm
```
2. Set the the ch number:
```
~/mcu/RIOT (2020.01-branch) $ git diff cpu/stm32_common/include/candev_stm32.h
diff --git a/cpu/stm32_common/include/candev_stm32.h b/cpu/stm32_common/include/candev_stm32.h
index 5a277c799076..3acd738431b8 100644
--- a/cpu/stm32_common/include/candev_stm32.h
+++ b/cpu/stm32_common/include/candev_stm32.h
@@ -41,7 +41,7 @@ extern "C" {
 #define CANDEV_STM32_CHAN_NUMOF 3
 #elif defined(CPU_FAM_STM32F1) || defined(CPU_FAM_STM32F2) || defined(CPU_FAM_STM32F4)
 #define CANDEV_STM32_CHAN_NUMOF 2
-#elif defined(CPU_FAM_STM32F0) || defined(CPU_FAM_STM32F3) || DOXYGEN
+#elif defined(CPU_FAM_STM32F0) || defined(CPU_FAM_STM32F3) || defined(CPU_FAM_STM32L4) || DOXYGEN
 /** Number of channels in the device (up to 3) */
 #define CANDEV_STM32_CHAN_NUMOF 1
```
3. Makefile:

```
## App name
#
APPLICATION = gtip-test-can

## Target
#
BOARD ?= nucleo-l476rg #nucleo-f413zh #nucleo-l476rg

## This has to be the absolute path to the RIOT base directory:
#
RIOTBASE ?= $(CURDIR)/../../RIOT/

## Chose the debug level: LOG_NONE, LOG_WARNING, LOG_ERROR, LOG_DEBUG, LOG_ALL
CFLAGS += -DLOG_LEVEL=LOG_DEBUG

CFLAGS += -DDEVELHELP

# Change this to 0 show compiler invocation lines by default:
QUIET ?= 1

# Modules to include:
USEMODULE += shell
USEMODULE += shell_commands
USEMODULE += ps

USEMODULE += conn_can
USEMODULE += can_isotp
USEMODULE += conn_can_isotp_multi
USEMODULE += can_pm
USEMODULE += can_trx

FEATURES_REQUIRED += periph_can
FEATURES_REQUIRED += periph_gpio_irq

CFLAGS += -DGNRC_PKTBUF_SIZE=4096
CFLAGS += -DCAN_PKT_BUF_SIZE=64
CFLAGS += -DCAN_ROUTER_MAX_FILTER=32

CFLAGS += -I$(CURDIR)
include $(RIOTBASE)/Makefile.include
```
4. Iusse log:
```
~/mcu/gtip-riotos/test_04 (master) $ make clean all
Building application "gtip-test-can" for "nucleo-l476rg" with MCU "stm32l4".

"make" -C /home/giuseppe/mcu/RIOT/boards/nucleo-l476rg
"make" -C /home/giuseppe/mcu/RIOT/boards/common/nucleo
"make" -C /home/giuseppe/mcu/RIOT/core
"make" -C /home/giuseppe/mcu/RIOT/cpu/stm32l4
"make" -C /home/giuseppe/mcu/RIOT/cpu/cortexm_common
"make" -C /home/giuseppe/mcu/RIOT/cpu/cortexm_common/periph
"make" -C /home/giuseppe/mcu/RIOT/cpu/stm32_common
"make" -C /home/giuseppe/mcu/RIOT/cpu/stm32_common/periph
"make" -C /home/giuseppe/mcu/RIOT/cpu/stm32l4/periph
"make" -C /home/giuseppe/mcu/RIOT/drivers
"make" -C /home/giuseppe/mcu/RIOT/drivers/can_trx
"make" -C /home/giuseppe/mcu/RIOT/drivers/periph_common
"make" -C /home/giuseppe/mcu/RIOT/sys
"make" -C /home/giuseppe/mcu/RIOT/sys/auto_init
"make" -C /home/giuseppe/mcu/RIOT/sys/auto_init/can
In file included from /home/giuseppe/mcu/RIOT/sys/auto_init/can/auto_init_periph_can.c:21:
/home/giuseppe/mcu/RIOT/cpu/stm32_common/include/can_params.h:42:21: error: 'RCC_APB1ENR_CAN1EN' undeclared here (not in a function); did you mean 'RCC_APB1ENR1_CAN1EN'?
   42 |         .rcc_mask = RCC_APB1ENR_CAN1EN,
      |                     ^~~~~~~~~~~~~~~~~~
      |                     RCC_APB1ENR1_CAN1EN
/home/giuseppe/mcu/RIOT/cpu/stm32_common/include/can_params.h:43:10: error: 'can_conf_t' {aka 'const struct <anonymous>'} has no member named 'can_master'
   43 |         .can_master = CAN1,
      |          ^~~~~~~~~~
In file included from /home/giuseppe/mcu/RIOT/cpu/stm32l4/include/cpu_conf.h:31,
                 from /home/giuseppe/mcu/RIOT/core/include/thread.h:125,
                 from /home/giuseppe/mcu/RIOT/sys/include/can/common.h:37,
                 from /home/giuseppe/mcu/RIOT/drivers/include/can/candev.h:35,
                 from /home/giuseppe/mcu/RIOT/sys/include/can/device.h:27,
                 from /home/giuseppe/mcu/RIOT/sys/auto_init/can/auto_init_periph_can.c:20:
/home/giuseppe/mcu/RIOT/cpu/stm32l4/include/vendor/stm32l476xx.h:1350:29: error: initialization of 'long unsigned int' from 'CAN_TypeDef *' {aka 'struct <anonymous> *'} makes integer from pointer without a cast [-Werror=int-conversion]
 1350 | #define CAN1                ((CAN_TypeDef *) CAN1_BASE)
      |                             ^
/home/giuseppe/mcu/RIOT/cpu/stm32_common/include/can_params.h:43:23: note: in expansion of macro 'CAN1'
   43 |         .can_master = CAN1,
      |                       ^~~~
/home/giuseppe/mcu/RIOT/cpu/stm32l4/include/vendor/stm32l476xx.h:1350:29: note: (near initialization for 'candev_conf[0].rx_pin')
 1350 | #define CAN1                ((CAN_TypeDef *) CAN1_BASE)
      |                             ^
/home/giuseppe/mcu/RIOT/cpu/stm32_common/include/can_params.h:43:23: note: in expansion of macro 'CAN1'
   43 |         .can_master = CAN1,
      |                       ^~~~
In file included from /home/giuseppe/mcu/RIOT/sys/auto_init/can/auto_init_periph_can.c:21:
/home/giuseppe/mcu/RIOT/cpu/stm32_common/include/can_params.h:44:10: error: 'can_conf_t' {aka 'const struct <anonymous>'} has no member named 'master_rcc_mask'
   44 |         .master_rcc_mask = RCC_APB1ENR_CAN1EN,
      |          ^~~~~~~~~~~~~~~
/home/giuseppe/mcu/RIOT/cpu/stm32_common/include/can_params.h:45:10: error: 'can_conf_t' {aka 'const struct <anonymous>'} has no member named 'first_filter'
   45 |         .first_filter = 0,
      |          ^~~~~~~~~~~~
/home/giuseppe/mcu/RIOT/cpu/stm32_common/include/can_params.h:46:10: error: 'can_conf_t' {aka 'const struct <anonymous>'} has no member named 'nb_filters'
   46 |         .nb_filters = 14,
      |          ^~~~~~~~~~
In file included from /home/giuseppe/mcu/RIOT/cpu/stm32l4/include/periph_cpu.h:23,
                 from /home/giuseppe/mcu/RIOT/boards/nucleo-l476rg/include/periph_conf.h:26,
                 from /home/giuseppe/mcu/RIOT/boards/common/nucleo/include/board_nucleo.h:25,
                 from /home/giuseppe/mcu/RIOT/boards/common/nucleo64/include/board.h:25,
                 from /home/giuseppe/mcu/RIOT/sys/include/xtimer.h:41,
                 from /home/giuseppe/mcu/RIOT/sys/include/can/device.h:31,
                 from /home/giuseppe/mcu/RIOT/sys/auto_init/can/auto_init_periph_can.c:20:
/home/giuseppe/mcu/RIOT/cpu/stm32_common/include/periph_cpu_common.h:161:29: error: initialized field overwritten [-Werror=override-init]
  161 | #define GPIO_PIN(x, y)      ((GPIOA_BASE + (x << 10)) | y)
      |                             ^
/home/giuseppe/mcu/RIOT/cpu/stm32_common/include/can_params.h:48:19: note: in expansion of macro 'GPIO_PIN'
   48 |         .rx_pin = GPIO_PIN(PORT_D, 0),
      |                   ^~~~~~~~
/home/giuseppe/mcu/RIOT/cpu/stm32_common/include/periph_cpu_common.h:161:29: note: (near initialization for 'candev_conf[0].rx_pin')
  161 | #define GPIO_PIN(x, y)      ((GPIOA_BASE + (x << 10)) | y)
      |                             ^
/home/giuseppe/mcu/RIOT/cpu/stm32_common/include/can_params.h:48:19: note: in expansion of macro 'GPIO_PIN'
   48 |         .rx_pin = GPIO_PIN(PORT_D, 0),
      |                   ^~~~~~~~
In file included from /home/giuseppe/mcu/RIOT/sys/auto_init/can/auto_init_periph_can.c:21:
/home/giuseppe/mcu/RIOT/cpu/stm32_common/include/can_params.h:50:15: error: initialized field overwritten [-Werror=override-init]
   50 |         .af = GPIO_AF9,
      |               ^~~~~~~~
/home/giuseppe/mcu/RIOT/cpu/stm32_common/include/can_params.h:50:15: note: (near initialization for 'candev_conf[0].af')
/home/giuseppe/mcu/RIOT/cpu/stm32_common/include/can_params.h:55:20: error: initialized field overwritten [-Werror=override-init]
   55 |         .tx_irqn = CAN1_TX_IRQn,
      |                    ^~~~~~~~~~~~
/home/giuseppe/mcu/RIOT/cpu/stm32_common/include/can_params.h:55:20: note: (near initialization for 'candev_conf[0].tx_irqn')
cc1: all warnings being treated as errors
make[4]: *** [/home/giuseppe/mcu/RIOT/Makefile.base:109: /home/giuseppe/mcu/gtip-riotos/test_04/bin/nucleo-l476rg/auto_init_can/auto_init_periph_can.o] Error 1
make[3]: *** [/home/giuseppe/mcu/RIOT/Makefile.base:29: ALL--/home/giuseppe/mcu/RIOT/sys/auto_init/can] Error 2
make[2]: *** [/home/giuseppe/mcu/RIOT/Makefile.base:29: ALL--/home/giuseppe/mcu/RIOT/sys/auto_init] Error 2
make[1]: *** [/home/giuseppe/mcu/RIOT/Makefile.base:29: ALL--/home/giuseppe/mcu/RIOT/sys] Error 2
make: *** [/home/giuseppe/mcu/gtip-riotos/test_04/../../RIOT//Makefile.include:540: /home/giuseppe/mcu/gtip-riotos/test_04/bin/nucleo-l476rg/application_gtip-test-can.a] Error 2
~/mcu/gtip-riotos/test_04 (master) $ 
```

5. Issues / missing feature 
    - the [candev_conf](https://github.com/RIOT-OS/RIOT/blob/master/cpu/stm32_common/include/can_params.h) missing the right setup for my target.
if my hypothesis is right I can try to fix it adding the support for stm32l4 target.

<!--
Example: Cannot build gnrc_networking application for samr21-xpro board.
-->

#### Steps to reproduce the issue
<!--
Try to describe as precisely as possible here the steps required to reproduce
the issue. Here you can also describe your hardware configuration, the network
setup, etc.
-->

#### Expected results
<!--
Example: The gnrc_networking application builds on samr21-xpro.
-->

#### Actual results
<!--
Please paste or specifically describe the actual output.
-->

#### Versions
<!--
Operating system: Mac OSX, Linux, Vagrant VM
Build environment: GCC, CLang versions (you can run the following command from
the RIOT base directory: make print-versions).
-->

<!-- Thanks for contributing! -->

--- a/boards/nucleo-l476rg/Makefile.features
+++ b/boards/nucleo-l476rg/Makefile.features
@@ -3,6 +3,7 @@ CPU_MODEL = stm32l476rg
 
 # Put defined MCU peripherals here (in alphabetical order)
 FEATURES_PROVIDED += periph_adc
+FEATURES_PROVIDED += periph_can
 FEATURES_PROVIDED += periph_dma
 FEATURES_PROVIDED += periph_i2c
 FEATURES_PROVIDED += periph_pwm

--- a/boards/nucleo-l476rg/doc.txt
+++ b/boards/nucleo-l476rg/doc.txt
@@ -6,23 +6,46 @@
 ### MCU
 
 
-| MCU        |       |
+| MCU        |   STM32L476RG    |
 |:------------- |:--------------------- |
 | Family | ARM Cortex-M4     |
 | Vendor | ST Microelectronics   |
-| RAM        |  |
-| Flash      | |
-| Frequency  | |
-| FPU        | |
-| Timers | |
-| ADCs       | |
-| UARTs      | |
-| SPIs       | |
-| I2Cs       | |
-| RTC        | |
-| Vcc        | |
+| RAM        | 128Kb |
+| Flash      | 1MB             |
+| Frequency  | up to 80MHz |
+| FPU        | yes               |
+| Timers | 16 (2x watchdog, 1 SysTick, 6x 16-bit, 2x 32-bit [TIM2])  |
+| ADCs       | 1x 12-bit         |
+| UARTs      | 3 (two UARTs and one Low-Power UART)                 |
+| SPIs       | 3                 |
+| I2Cs       | 3                 |
+| RTC        | 1                 |
+| CAN        | 1                 |
+| Vcc        | 1.71 V - 3.6V           |
 | Datasheet  | |
 | Reference Manual | [Reference Manual](http://www.st.com/content/ccc/resource/technical/document/reference_manual/02/35/09/0c/4f/f7/40/03/DM00083560.pdf/files/DM00083560.pdf/jcr:content/translations/en.DM00083560.pdf) |
 | Programming Manual | [Programming Manual](http://www.st.com/content/ccc/resource/technical/document/programming_manual/6c/3a/cb/e7/e4/ea/44/9b/DM00046982.pdf/files/DM00046982.pdf/jcr:content/translations/en.DM00046982.pdf) |
-| Board Manual   | |
+| Board Manual   | [Board Manual](https://www.st.com/content/ccc/resource/technical/document/user_manual/98/2e/fa/4b/e0/82/43/b7/DM00105823.pdf/files/DM00105823.pdf/jcr:content/translations/en.DM00105823.pdf) |
+
+## Flashing the device
+
+The ST Nucleo-L476RG board includes an on-board ST-LINK V2 programmer. The
+easiest way to program the board is to use OpenOCD. Once you have installed
+OpenOCD (look [here](https://github.com/RIOT-OS/RIOT/wiki/OpenOCD) for
+installation instructions), you can flash the board simply by typing
+
+```
+make BOARD=nucleo-l476rg flash
+```
+and debug via GDB by simply typing
+```
+make BOARD=nucleo-l476rg debug
+```
+
+
+## Supported Toolchains
+
+For using the ST Nucleo-L476RG board we strongly recommend the usage of the
+[GNU Tools for ARM Embedded Processors](https://launchpad.net/gcc-arm-embedded)
+toolchain.
  */

--- a/cpu/stm32_common/include/can_params.h
+++ b/cpu/stm32_common/include/can_params.h
@@ -39,18 +39,26 @@ static const can_conf_t candev_conf[] = {
         .irqn = CEC_CAN_IRQn,
 #else
         .can = CAN1,
+#if defined(CPU_FAM_STM32L4)
+        .rcc_mask = RCC_APB1ENR1_CAN1EN,
+#else
         .rcc_mask = RCC_APB1ENR_CAN1EN,
         .can_master = CAN1,
         .master_rcc_mask = RCC_APB1ENR_CAN1EN,
         .first_filter = 0,
         .nb_filters = 14,
-#ifndef CPU_FAM_STM32F1
+#endif
+#if  defined(CPU_FAM_STM32F1)
+        .rx_pin = GPIO_PIN(PORT_A, 11),
+        .tx_pin = GPIO_PIN(PORT_A, 12),
+#elif defined(CPU_FAM_STM32L4)
+        .rx_pin = GPIO_PIN(PORT_B, 8),
+        .tx_pin = GPIO_PIN(PORT_B, 9),
+        .af = GPIO_AF9,
+#else
         .rx_pin = GPIO_PIN(PORT_D, 0),
         .tx_pin = GPIO_PIN(PORT_D, 1),
         .af = GPIO_AF9,
-#else
-        .rx_pin = GPIO_PIN(PORT_A, 11),
-        .tx_pin = GPIO_PIN(PORT_A, 12),
 #endif
         .tx_irqn = CAN1_TX_IRQn,
         .rx0_irqn = CAN1_RX0_IRQn,

--- a/cpu/stm32_common/include/candev_stm32.h
+++ b/cpu/stm32_common/include/candev_stm32.h
@@ -41,7 +41,8 @@ extern "C" {
 #define CANDEV_STM32_CHAN_NUMOF 3
 #elif defined(CPU_FAM_STM32F1) || defined(CPU_FAM_STM32F2) || defined(CPU_FAM_STM32F4)
 #define CANDEV_STM32_CHAN_NUMOF 2
-#elif defined(CPU_FAM_STM32F0) || defined(CPU_FAM_STM32F3) || DOXYGEN
+#elif defined(CPU_FAM_STM32F0) || defined(CPU_FAM_STM32F3) || \
+      defined(CPU_FAM_STM32L4) || DOXYGEN
 /** Number of channels in the device (up to 3) */
 #define CANDEV_STM32_CHAN_NUMOF 1
 #else

